compiler/rustc_index/src/vec.rs

```diff
@@ -12,7 +12,19 @@ use std::vec;
 use crate::{Idx, IndexSlice};
 
 /// An owned contiguous collection of `T`s, indexed by `I` rather than by `usize`.
-/// Its purpose is to avoid mixing indexes.
+///
+/// Why use this instead of a `Vec`?
+/// This enforces the user to index a given IndexVec only with the associated index thus making it
+/// impossible to use the wrong index for a given `IndexVec`.
```

Comment:
Might be good to clarify that index is a specific type, something like

```suggestion
/// An `IndexVec` allows element access only via a specific associated index type, meaning that
/// trying to use the wrong index type (possibly accessing an invalid element) will fail at
/// compile time.
```

Also the ``Why use this instead of a `Vec`?`` line above will just get smooshed into this line (add an empty line if you want it in a separate paragraph)

Reply:
Oh I didn't realize.